api/v1beta1/dbinstance_types.go

```diff
@@ -0,0 +1,194 @@
+}
+
+// BackendServer defines backend database server
+type BackendServer struct {
+	Host          string `json:"host"`
+	Port          uint16 `json:"port"`
+	MaxConnection uint16 `json:"maxConn"`
+	ReadOnly      bool   `json:"readonly,omitempty"`
+}
+
+// GenericInstance is used when instance type is generic
+// and describes necessary informations to use instance
+// generic instance can be any backend, it must be reachable by described address and port
+type GenericInstance struct {
+	Host     string `json:"host"`
+	Port     uint16 `json:"port"`
+	PublicIP string `json:"publicIp,omitempty"`
+	// BackupHost address will be used for dumping database for backup
+	// Usually secondary address for primary-secondary setup or cluster lb address
+	// If it's not defined, above Host will be used as backup host address.
+	BackupHost string `json:"backupHost,omitempty"`
+}
+
+// DbInstanceBackup defines name of google bucket to use for storing database dumps for backup when backup is enabled
+type DbInstanceBackup struct {
+	Bucket string `json:"bucket"`
+}
+
+// DbInstanceMonitoring defines if exporter
+type DbInstanceMonitoring struct {
+	Enabled bool `json:"enabled"`
+}
+
+// DbInstanceSSLConnection defines weather connection from db-operator to instance has to be ssl or not
+type DbInstanceSSLConnection struct {
+	Enabled bool `json:"enabled"`
+	// SkipVerity use SSL connection, but don't check against a CA
+	SkipVerify bool `json:"skip-verify"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+//+kubebuilder:resource:scope=Cluster,shortName=dbin
+//+kubebuilder:printcolumn:name="Phase",type=string,JSONPath=`.status.phase`,description="current phase"
+//+kubebuilder:printcolumn:name="Status",type=string,JSONPath=`.status.status`,description="health status"
+// +kubebuilder:storageversion
+
+// DbInstance is the Schema for the dbinstances API
+type DbInstance struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   DbInstanceSpec   `json:"spec,omitempty"`
+	Status DbInstanceStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// DbInstanceList contains a list of DbInstance
+type DbInstanceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []DbInstance `json:"items"`
+}
+
+func init() {
+	SchemeBuilder.Register(&DbInstance{}, &DbInstanceList{})
+}
+
+// ValidateEngine checks if defined engine by DbInstance object is supported by db-operator
+func (dbin *DbInstance) ValidateEngine() error {
+	if (dbin.Spec.Engine == "mysql") || (dbin.Spec.Engine == "postgres") {
+		return nil
+	}
+
+	return errors.New("not supported engine type")
+}
+
+// ValidateBackend checks if backend type of instance is defined properly
+// returns error when more than one backend types are defined
+// or when no backend type is defined
+func (dbin *DbInstance) ValidateBackend() error {
```

Comment:
Wondering if this should maybe go in future into the webhook validation? But maybe a future issue.

Reply:
I would try, but rather as a separated issue